### DISCUSSION TITLE
Emscripten wasm doesn't work as in ObservableHQ site

### DISCUSSION THREAD

**Discussion by gganapat** [observablehq/framework#1033] [Q&A] [answered]
I am not able to get wasm Module available for use as could be done on the web version.

```js
const url = "https://coolprop-gani.s3.us-west-2.amazonaws.com/coolprop.js"
```

```js
const Module = fetch(url).catch(() => {
  const m = window.Module;
  if (m.onRuntimeInitialized) return m; // Already initialized.
  return new Promise(resolve => m.onRuntimeInitialized = () => resolve(m));
})
```
```js
display(Module)
```
<img width="438" alt="Screenshot 2024-03-09 at 10 24 56 AM" src="https://github.com/observablehq/framework/assets/46061218/1101697a-d85e-4bc5-9fb0-146b2d9339f3">

Not sure how to proceed. Help wanted.

Thanks,
Gani --

**Comment by mbostock** [comment]
Did you mean to call import instead of fetch?

**Comment by gganapat** [comment]
Tried that but it didn't work.

```js
const Module = import(url).catch(() => {
  const m = window.Module;
  if (m.onRuntimeInitialized) return m; // Already initialized.
  return new Promise(resolve => m.onRuntimeInitialized = () => resolve(m));
})
```
```js
display(Module)
```
```js
display(Module.PropsSI('D','P',1000000.,'T',273.15-20,'Xenon'))
```

<img width="429" alt="Screenshot 2024-03-09 at 4 20 26 PM" src="https://github.com/observablehq/framework/assets/46061218/c2347e06-9476-437d-b34e-5bc64fe8f7a5">

**Comment by mbostock** [comment]
You probably need to follow the documentation for the library you’re trying to load, which I’m guessing is this:

http://www.coolprop.org/coolprop/wrappers/Javascript/index.html

I’m converting this to a discussion since this isn’t a bug in Framework.

**Comment by gganapat** [comment]
Thanks! But it works fine in ObservableHQ - see [link](https://observablehq.com/@gganapat/coolprop-wasm-exercise).

I'm hoping there is a workaround.

Thanks,
Gani --

**Comment by mbostock** [comment, marked as answer]
Observable Framework is just vanilla JavaScript and HTML, so you can just add a script tag as shown in the documentation:

```md
<script src="https://coolprop-gani.s3.us-west-2.amazonaws.com/coolprop.js"></script>
```

If you then want to wait for the module to load, it looks like you can do something like this:

````md
```js
if (!Module.onRuntimeInitialized) await new Promise((resolve) => (Module.onRuntimeInitialized = resolve));
const coolprop = Module;
```
````

Here’s a minimal example of using it:

````md
```js
const fluid = "Xenon";
const temp = 25;
const psia_to_Pas = 6894.75729;
```

```js
Plot.plot({
  title: `Density of ${fluid} at ${temp}°C`,
  x: {label: "Pressure (psia)"},
  y: {label: "Density (kg/m³)"},
  marks: [
    Plot.ruleY([0]),
    Plot.lineY(d3.ticks(0, 1200, 100), {
      x: (p) => p,
      y: (p) => coolprop.PropsSI("Dmass", "T", temp + 273.15, "P", p * psia_to_Pas, fluid),
      tip: true
    })
  ]
})
```
````

The chart looks like this:

<img width="667" alt="Screenshot 2024-03-09 at 9 12 17 PM" src="https://github.com/observablehq/framework/assets/230541/d6e42a2f-599c-45e0-90f5-fac4f8c47388">

But probably you want to download coolprop.js and its WebAssembly bundle (.wasm) so that it’s self-hosted within your Framework project rather than loading it from AWS. And ideally someone would rebuild so that it’s a modern ES module that you can import rather than needing a vanilla script tag and the `Module.onRuntimeInitialized` business.

**Reply by gganapat** [reply]
ok, fantastic! That worked beautifully. I wonder why the code from ObservableHQ webapp was not directly transferable to Framework.

Thanks!
Gani -

**Comment by gganapat** [comment]
I followed up on your suggestion to download coolprop.js and coolprop.wasm, and tried several things to make it work. I tried your solution in [here](https://talk.observablehq.com/t/how-to-load-a-local-js-file-into-observable/3219/6) and [here](https://observablehq.com/@circles/load-a-library-from-a-file-attachment) and other versions but none of them worked. I spent a morning trying to get it work but no success. If it's something that I missed and you won't have to spend much time trouble-shooting it, I would appreciate your help.

Thanks,
Gani --

**Reply by gganapat** [reply]
For example when I did this, 

```js
const t = (await import(await FileAttachment("coolprop.js").url())).SCRIPTS
```

```js
if (!Module.onRuntimeInitialized) await new Promise((resolve) => (Module.onRuntimeInitialized = resolve));
const coolprop = Module;
display(coolprop)
```
I get this ...

<img width="1010" alt="Screenshot 2024-03-10 at 4 58 18 PM" src="https://github.com/observablehq/framework/assets/46061218/21f44678-36f9-4c41-a2fb-8b313dbe051a">

Any insight would help!

Thanks, Gani -

**Reply by mbostock** [reply]
It won’t work in Framework the way CoolProp is currently built. You’ll have to host it elsewhere, or you’ll need to edit the CoolProp source.

The problem is that `coolprop.js` hard-codes the relative path to `coolprop.wasm` and doesn’t let you override it. That won’t work in Framework because we use content-hashed file names for all assets. (If we add support for a public folder #169 maybe we could allow this… but even then it’s not clear to me how you’d load a script from the public folder, as normally scripts are treated as assets and given content-hashed file names in `_file`.)

Modern libraries use `import.meta.resolve` instead of hard-coding a path, and Framework supports rewrites these as needed. Some WebAssembly libraries also let you pass in the path to the `.wasm` file, but CoolProp doesn’t appear to let you do that, so I don’t see a way of fixing it (short of editing the code manually to add such a feature).

Also, it seems like my code above has a race condition where sometimes it doesn’t receive the `onRuntimeInitialized` callback; like I said above, there are better ways to build this WebAssembly library using modern JavaScript and ES modules that would make this work out of the box. But I got it working like so:

````md
<script>

var coolprop = new Promise((resolve) => {
  Module = {
    onRuntimeInitialized() {
      resolve(Module);
    }
  };
});

</script>
<script src="coolprop.js"></script>
````

This mess is because CoolProp is packaged as a global script rather than a modern ES module; in general it’s not good to use `<script src>` because it is synchronous, and likewise it’s messy to use globals such as `Module`.

Also, there’s a bug where you can’t use `<script src>` in Framework for a local script. I opened https://github.com/observablehq/framework/pull/1035 to fix that. But even if we fix that, you’ll have the `coolprop.wasm` issue described above.

**Reply by gganapat** [reply]
Thanks so much! So, you were able to run the code as above? I was not able to - I got this...

<img width="1175" alt="Screenshot 2024-03-10 at 6 21 46 PM" src="https://github.com/observablehq/framework/assets/46061218/01a624b9-d99f-4b68-9cfb-077c5677bf31">

**Reply by mbostock** [reply]
I was in the PR, but like I said, it still won’t work on build because of the aforementioned limitations.

**Reply by gganapat** [reply]
Ok, thanks for all your help! You have been amazing.

gani-
